### PR TITLE
deprecate constraints in favor of dynamic_shapes

### PR DESCRIPTION
**Pull request by avikchaudhuri**
Recently we updated the `export` API to take an experimental `dynamic_shapes` argument that was meant to subsume the existing `constraints` argument. 

This PR deprecates `constraints` (with a warning on its use, but without actually removing it). Simultaneously it replaces all uses of `constraints` in docs, examples, and tests with corresponding uses of `dynamic_shapes` (preserving behavior). This exercise fortunately revealed some minor bugs in the implementation which have also been fixed in this PR.

Some uses of `constraints` still remain, e.g., when `torch._dynamo.export` is called directly. (Meta-internal uses will be updated in a separate diff.)

Differential Revision: D49676049

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by tugsbayasgalan — APPROVED** [review]

File: test/dynamo/test_export.py [hunk 1]
```diff
@@ -2238,17 +2238,16 @@ def foo(x):
                 return x + 2
 
         t = torch.zeros([8, 4])
-        constraints = [
```
Comment (gmagogsfm) [inline test/dynamo/test_export.py:2241]: Can we add additional unit tests instead of changing existing ones. Though `Dim` API eventually also constructs `Constraint`s, there are still logic covered in current unit test (like dynamic_dim(x,0)>=3) that is not covered by `Dim`

Comment (avikchaudhuri) [reply test/dynamo/test_export.py:2241]: Not sure what you mean, `Dim` does have lower / upper bounds and the modified test preserves the exact behavior of the original test.

That said, there are a lot of direct unit tests for `dynamic_dim` that use `torch._dynamo.export`, which continues to use constraints. So the existing functionality of `dynamic_dim` (bounds, equalities) remains well-tested.

My purpose in modifying tests is two-fold: (a) make sure `Dim` has good coverage (b) avoid warnings in tests, which is a strange thing to have. 

Finally, once we decide to remove `constraints` I would like to remove `dynamic_dim` entirely, since it is way more direct to go from `Dim` to symbols instead of going through `dynamic_dim`. So this work needed to be done anyway.
